File: fm-domain/src/main/java/io/zeitwert/dddrive/ddd/adapter/api/jsonapi/dto/AggregateDtoAdapterBase.java
package io.zeitwert.dddrive.ddd.adapter.api.jsonapi.dto;

import dddrive.app.doc.model.Doc;
import dddrive.app.obj.model.Obj;
import dddrive.ddd.core.model.Aggregate;
import io.zeitwert.dddrive.ddd.adapter.api.jsonapi.AggregateDto;
import io.zeitwert.dddrive.ddd.adapter.api.jsonapi.AggregateDtoAdapter;
import io.zeitwert.fm.oe.model.ObjTenant;
import io.zeitwert.fm.oe.model.ObjTenantRepository;
import io.zeitwert.fm.oe.model.ObjUser;
import io.zeitwert.fm.oe.model.ObjUserRepository;
import org.springframework.beans.factory.annotation.Autowired;

public abstract class AggregateDtoAdapterBase<A extends Aggregate, D extends AggregateDto<A>>
		implements AggregateDtoAdapter<A, D> {

	private ObjTenantRepository tenantRepository = null;
	private ObjUserRepository userRepository = null;

	protected ObjTenantRepository getTenantRepository() {
		return this.tenantRepository;
	}

	@Autowired
	void setTenantRepository(ObjTenantRepository tenantRepository) {
		this.tenantRepository = tenantRepository;
	}

	protected ObjTenant getTenant(Object tenantId) {
		return this.tenantRepository.get(tenantId);
	}

	protected EnumeratedDto getTenantEnumerated(Integer tenantId) {
		return tenantId != null ? EnumeratedDto.of(tenantRepository.get(tenantId)) : null;
	}

	protected ObjUserRepository getUserRepository() {
		return this.userRepository;
	}

	@Autowired
	void setUserRepository(ObjUserRepository userRepository) {
		this.userRepository = userRepository;
	}

	protected ObjUser getUser(Object userId) {
		return this.userRepository.get(userId);
	}

	protected EnumeratedDto getUserEnumerated(Integer userId) {
		return userId != null ? EnumeratedDto.of(userRepository.get(userId)) : null;
	}

	protected EnumeratedDto asEnumerated(Aggregate a) {
		if (a instanceof Obj) {
			return asEnumerated((Obj) a);
		} else {
			return asEnumerated((Doc) a);
		}
	}

	protected EnumeratedDto asEnumerated(Obj o) {
		return EnumeratedDto.of(o.getId().toString(), o.getCaption());
	}

	protected EnumeratedDto asEnumerated(Doc d) {
		return EnumeratedDto.of(d.getId().toString(), d.getCaption());
	}

}

File: fm-domain/src/main/java/io/zeitwert/dddrive/ddd/api/rest/base/AggregateDtoAdapterBase.kt
package io.zeitwert.dddrive.ddd.api.rest.base

import dddrive.ddd.core.model.Aggregate
import dddrive.ddd.core.model.AggregateRepository
import io.zeitwert.dddrive.ddd.api.rest.AggregateDtoAdapter
import io.zeitwert.dddrive.ddd.api.rest.DtoDetailLevel
import io.zeitwert.dddrive.ddd.api.rest.DtoUtils
import io.zeitwert.dddrive.ddd.api.rest.dto.AggregateDto
import io.zeitwert.fm.oe.model.ObjTenant
import io.zeitwert.fm.oe.model.ObjUser
import org.springframework.beans.factory.annotation.Autowired

abstract class AggregateDtoAdapterBase<A : Aggregate, D : AggregateDto<A>>(
	protected val repo: AggregateRepository<A>,
) : AggregateDtoAdapter<A, D> {

	private lateinit var tenantRepo: AggregateRepository<ObjTenant>
	private lateinit var userRepo: AggregateRepository<ObjUser>

	override fun fromAggregate(
		id: String,
		detailLevel: DtoDetailLevel,
	): D = fromAggregate(repo.get(DtoUtils.idFromString(id)!!), detailLevel)

	abstract override fun fromAggregate(
		aggregate: A,
		detailLevel: DtoDetailLevel,
	): D

	@Autowired
	protected fun setTenantRepo(tenantRepo: AggregateRepository<ObjTenant>) {
		this.tenantRepo = tenantRepo
	}

	protected fun getTenant(tenantId: Any): ObjTenant = tenantRepo.get(tenantId)

	@Autowired
	protected fun setUserRepo(userRepo: AggregateRepository<ObjUser>) {
		this.userRepo = userRepo
	}

	protected fun getUser(userId: Any): ObjUser = userRepo.get(userId)

}
